### PR TITLE
Pmap error

### LINKED ISSUE #1991: pmap has a bad error message for when the leading axis doesn't exist.
This code

```python
def f(a, b):
  for i in range(b):
    a += a
  return a

jax.pmap(f)(jnp.ones((8, 100)), 10)
```

Gives this error.
```
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-16-092495ffbdde> in <module>()
----> 1 jax.pmap(f)(jnp.ones((8, 100)), 10)

6 frames
jax/interpreters/pxla.py in _shard_abstract_array(size, x)
    132 shard_aval_handlers[core.AbstractUnit] = lambda size, x: x
    133 def _shard_abstract_array(size, x):
--> 134   if x.shape[0] != size:
    135     raise ValueError("Axis size {} does not match leading dimension of "
    136                      "shape {}".format(size, x.shape))

IndexError: tuple index out of range
```

This is happening because the `10` isn't a vector. (The correct way to call the last line should be
`jax.pmap(partial(f, b=10))(jnp.ones((8, 100)))`)

However, this error is really bad, and I only knew how to fix it because I've hit it many times before. We should explicitly check for this.

### PR DESCRIPTION
**Pull request by chaserileyroberts**
Fixes #1991.

### CODE REVIEW

File: tests/pmap_test.py [hunk 1]
```diff
@@ -90,6 +90,16 @@ def testBasic(self):
...
 
+  def testNoLeadingAxis(self):
+    f = pmap(lambda x, y: x + y, axis_name='i')
+    shape = (xla_bridge.device_count(), 4)
+    x = onp.arange(prod(shape), dtype=onp.float32).reshape(shape)
+    ans = lambda: f(x, 1)
+    self.assertRaisesRegex(
+        ValueError,
+        "Some argument is missing a leading dimesion. (Did you mean to use a partial?)"
+        ans)
```
Comment (skye) [inline tests/pmap_test.py]: This is causing a syntax error, I think you need a comma on the above line.

**Comment by skye** [comment]
This is definitely an improvement, thanks! I would go even farther and plumb in some more debugging information to make this more actionable, e.g. which argument index is missing a leading axis? I'm happy to submit this as-is and iterate as well, lemme know what you wanna do.

**Comment by Thenerdstation** [comment]
> ! I would go even farther and plumb in some more debugging information to make this more actionable, e.g. which argument index is missing a leading axis?

That's a good idea, though I'm not sure the cleanest way of doing that. Ideally, it would point at the exact `pmap` line that failed, but I'm not sure how.

Something like:
```
jax.pmap(f)(jnp.ones((8, 100)), 10)
                                ^
ValueError: Argument #2 is missing a leading axis
```
?

**Comment by skye** [comment]
I think you need to plumb the argument indices through into `shard_args` here: https://github.com/google/jax/blob/master/jax/interpreters/pxla.py#L456 (i.e. map over `enumerate(avals)` instead of `avals`, and then plumb the index through to `shard_args` and the underlying handlers). That should let you specify the argument index in the error message, and the stack trace will contain the failing pmap line, which is pretty close!

**Comment by hawkinsp** [comment]
I believe the underlying issue is fixed (see the linked issue.)

**Pull request closed without merging**